File: elgamal.cpp
/*
    CPP
*/

#include <iostream>

/*
    For eficiency module and power (pow)
*/
int powMod(int base, int root, int mod) {
    int dummy = base;

    for (int i = 0; i < root-1; i++) {
        dummy = (dummy * base) % mod;
    }

    return dummy;
}

int isRelativePrime(int a, int b) {
    int dummy = 0;

    while (b != 0) {
        dummy = a % b;
        a = b;
        b = dummy;
    }

    return a;    
}

int isPrimitiveRoot(int prime, int root) {
    int h[prime-2];
    int hprime = prime;

    for(int i=0; i<prime; i++) {
        h[i] = powMod(root, i+1, prime);
    }
    
    int kampret = 1;
    int arrsize = sizeof(h)/sizeof(h[0]);
    for(int i=0; i<arrsize-1; i++) {
        for (int j=kampret; j<arrsize; j++) {
            if (h[i] == h[j]) {
                std::cout << "Not primite root" << std::endl;
                return 0;
            }
        }
        
        if(isRelativePrime(prime, h[i]) != 1) {
            return 0;
        }

        kampret++;
    }

    return 1;
}

void keyBuilder(int p, int g, int x) {
    int y;
    if(isPrimitiveRoot(p, g) == 1) {
        y = powMod(g, x, p);
        std::cout << "Public key = (" << y << ", " << g << ", " << p << ")" << std::endl; 
        std::cout << "Private key = " << x << std::endl; 
    } else {
        std::cout << g << " is not primitive root from " << p << std::endl;
    }
}

void encryption() {
    int msg = 48, y = 0, g = 0, p = 0, x = 0;
    std::cout << "Input y, g, p = ";
    std::cin >> y >> g >> p;
    
    std::cout << "Input k (1 < k < p-1) = ";
    int k = 0;
    std::cin >> k;
    
    int a = powMod(g, k, p);
    int b = powMod(y, k, p) * msg % p;

    std::cout << "Ciphertext = (" << a << ", " << b << ")" << std::endl;
}

void decryption() {
    int a, b, x, p, m;
    std::cout << "Input a, b, x, p = ";
    std::cin >> a >> b >> x >> p;
    m = powMod(a, p-1-x, p) * b % p;
    std::cout << "Message = " << m << std::endl;
}

int main() {
    int p = 0, g = 0, x = 0;
    std::cin >> p >> g >> x;
    keyBuilder(p, g, x);

    encryption();
    decryption();
}

File: elgamal.py
def is_relative_prime(prime, param):
    while param != 0:
        prime, param = param, prime % param
    return prime

def is_primitive_root(prime, root):
    """
    Root is primitive root from prime
    :param prime: int
    :param root: int
    :return: int
    """
    primitive_root = []
    for i in range(prime - 1):
        primitive_root.append(pow(root, i + 1) % prime)

    # Untuk memastikan tidak terjadi pembandingan primitive_root pada indeks j dan i yang sama
    kampret = 1
    
    # Membandingkan apakah tidak ada data yang sama
    for i in range(len(primitive_root)):
        for j in range(kampret, len(primitive_root) - 1):
            if primitive_root[i] == primitive_root[j + 1]:
                print("Not primitive root")
                return 0

        if is_relative_prime(prime, primitive_root[i]) != 1:
            print("Not relative prime")
            return 0
        kampret += 1
    return 1


def key_builder():
    print("Input p, g and x")
    prime, root, secret_key = [int(x) for x in input().split(" ")]
    if is_primitive_root(prime, root) == 1:
        y = pow(root, secret_key) % prime
        print(f"Public Key (%d, %d, %d)" % (y, root, prime))
        print(f"Private key %d" % secret_key)


def encryption():
    print("Input message")
    message = input()

    print("Input y, g and p (public key)")
    y, g, p = [int(x) for x in input().split(" ")]

    print("Input k")
    k = int(input())

    a = pow(g, k) % p
    b = (pow(y, k) * ord(message)) % p

    print(f"Ciphertext (%d, %d)" % (a, b))


def decryption():
    a, b, x, p = [int(x) for x in input().split(" ")]
    m = pow(a, (p-1-x)) * b % p
    print(f"Message %c" % m)


key_builder()
# 2273 3 243
# Public Key 461 3 2273
# Private key 243
encryption()
# Input k
# 1463
# Ciphertext (1439, 2004)
decryption()